### PR TITLE
[rebased/cherry-picked] Switch to JarsToLabels provider and rework/cleanup scala_import

### PR DESCRIPTION
**Pull request by andyscott**
This is a rebased version of #477 plus the addition of @jjudd's changes in https://github.com/andyscott/rules_scala/pull/2.

### CODE REVIEW

**Comment by andyscott** [comment]
I goofed-- this will need some more work to make sure all the changes were pulled over correctly.

**Comment by ittaiz** [comment]
Is this really a good idea to pile up a few different changes on a big
refactoring?
It makes reviewing much harder.
I’m still not sure about the core change and exports...

**Comment by andyscott** [comment]
I'm not too attached to this work, but I'll have to defer to @jjudd since he added the more recent changes to make the original refactoring support IntelliJ. Also, given some recent discussions, I wanted to make sure we had a more manageable branch based on recent refactoring.

**Comment by jjudd** [comment]
We are running on a branch that has all these changes on it, and things have been working fine. I'm not too attached to this work either, as we'll be switching to Annex internally this week for the Zinc support.

**Comment by jjudd** [comment]
Also, I consent to my changes being contributed.

File: scala/private/rule_impls.bzl [hunk 1]
```diff
@@ -122,12 +122,14 @@ def _collect_plugin_paths(plugins):
...
+        elif p[JavaInfo] and p[JavaInfo].full_compile_jars:
+            paths.extend(p[JavaInfo].full_compile_jars.to_list())
         elif hasattr(p, "scala"):
             paths.append(p.scala.outputs.jar)
         elif hasattr(p, "java"):
             paths.extend([j.class_jar for j in p.java.outputs.jars])
-        # support http_file pointed at a jar. http_jar uses ijar,
-        # which breaks scala macros
+            # support http_file pointed at a jar. http_jar uses ijar,
+            # which breaks scala macros
```
Comment (johnynek) [inline scala/private/rule_impls.bzl]: this comment is on the elif, can we move it below or unindent? It feels awkward to me being indented but not related to the current block.

File: scala/providers.bzl [hunk 2]
```diff
@@ -30,3 +30,10 @@ def create_scala_provider(
         transitive_runtime_jars = transitive_runtime_jars,
         transitive_exports = [] #needed by intellij plugin
     )
+
+JarsToLabels = provider(
```
Comment (johnynek) [inline scala/providers.bzl]: can we make a link to the bazel discussions about putting labels in the jars?

https://github.com/bazelbuild/bazel/issues/4584

**Comment by ittaiz** [comment]
I agree but also introducing bugs in a critical flow would be unfortunate.
Andi,
Did you get a chance to look at the exports thing?

**Pull request closed without merging**